Fixed latex output for multi-indexed dataframes - GH9778

yred opened the pull request:
Proposed fix for https://github.com/pydata/pandas/issues/9778

The formatting issue was caused by an incorrect number of elements in the (first) index columns of `strcols`. The length of reinserted columns was based on the number of elements per index level, but should have relied on the number of rows/occurrences of such elements.

yred committed:
BUG: Fixed latex output for multi-indexed dataframes - GH9778

yred commented on pandas/core/format.py:
```diff
@@ -613,8 +613,12 @@ def get_col_type(dtype):
+                lev3 = [blank] * clevels
+                for level_idx, group in itertools.groupby(
+                        self.frame.index.labels[i]):
+                    count = sum(1 for _ in group)
```
I've originally had this as:

```python
count = len(list(group))
```

but was wondering if using `sum()` makes a bit more sense from a performance perspective.

shoyer replied:
Performance shouldn't really be a concern here... nobody is going to output latex for tables much larger than can fit on a single page.

yred replied:
Would it then make more sense to switch it back to how it was? (I guess the first form is a bit more readable/intuitive).

shoyer replied:
doesn't really matter to me either way :)

yred replied:
Changed it back to the original form (which seems unexpectedly faster from a simple test).

shoyer replied:
In general, you shouldn't trust benchmarks that show list expressions are faster than generators... allocating memory in repeatedly in a benchmark is faster than it is in real use. But again, this is not performance limited code.

shoyer commented:
This looks great to me. Can you please:

1. [add a note to what's new](http://pandas-docs.github.io/pandas-docs-travis/contributing.html#documenting-your-code)
2. [squash your changes into one commit](http://pandas-docs.github.io/pandas-docs-travis/contributing.html#committing-your-code) (we prefer not commits for which Travis tests fail merged into master, because it can complicate things like `git bisect`)

yred commented:
@shoyer: Thanks a lot for your feedback.

I've just added a note to the next release, and squashed all the commits. Do let me know if there's anything else that should be added/updated.

shoyer commented:
This looks to good to go, once the tests on Travis pass. I'll try to check later, but please feel free to ping me if you notice that's happened.

yred commented:
@shoyer: I'm not sure how Travis is set up on this repo, but [tests are passing](https://travis-ci.org/yred/pandas) on my fork.

hayd commented:
Do we handle if there are `df.index.names` ?

Thanks for fixing this!

yred commented:
@hayd: That's a good point.

It currently doesn't print out index names. Would that be expected with this fix or should I add it in a separate PR?

shoyer commented:
At this point probably better to make a separate PR. But I do recall being able to export index names....

On Wed, Apr 15, 2015 at 6:11 PM, Yasin A. <notifications@github.com>
wrote:

> @hayd: That's a good point.
> It currently doesn't print out index names. Would that be expected with this fix or should I add it in a separate PR?
> ---
> Reply to this email directly or view it on GitHub:
> https://github.com/pydata/pandas/pull/9908#issuecomment-93610178

yred commented:
@shoyer: it seems that regression happened a little earlier as well.

I'll be updating the code for that, and potentially cleaning up `to_latex()` so that things are a bit clearer/cleaner.

hayd commented:
@yred Fantastic, thanks!

shoyer commented:
thanks @yred !
